packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,664 @@
+  }
+
+  updateCaveat(
+    origin: string,
+    target: string,
+    caveatType: string,
+    caveatValue: Json,
+  ): void {
+    if (!this.hasCaveat(origin, target, caveatType)) {
+      throw new CaveatDoesNotExistError(origin, target, caveatType);
+    }
+
+    this.setCaveat(origin, target, caveatType, caveatValue);
+  }
+
+  private getCaveat(
+    origin: string,
+    target: string,
+    caveatType: string,
+  ): Caveat<Json> | undefined {
+    const permission = this.getPermission(origin, target);
+    if (!permission) {
+      throw new PermissionDoesNotExistError(origin, target);
+    }
+
+    return findCaveat(permission, caveatType);
+  }
+
+  private setCaveat(
+    origin: string,
+    target: string,
+    caveatType: string,
+    caveatValue: Json,
+  ): void {
+    this.update((draftState) => {
+      const permission = draftState.subjects[origin]?.permissions[target];
+      /* istanbul ignore if: This should be impossible, but TypeScript wants it */
+      if (!permission) {
+        throw new PermissionDoesNotExistError(origin, target);
+      }
+
+      const caveat = this.computeCaveat(origin, target, {
+        type: caveatType,
+        value: caveatValue,
+      });
+
+      // Typecasts: ts(2589)
+      if (permission.caveats) {
+        const caveatIndex = permission.caveats.findIndex(
+          (existingCaveat) => existingCaveat.type === caveat.type,
+        );
+
+        if (caveatIndex === -1) {
+          permission.caveats.push(caveat as any);
+        } else {
+          permission.caveats.splice(caveatIndex, 1, caveat as any);
+        }
+      } else {
+        permission.caveats = [caveat as any];
+      }
+
+      // Typecast: ts(2589)
+      this.validatePermission(permission as any);
+    });
+  }
+
+  removeCaveat(origin: string, target: string, caveatType: string): void {
+    this.update((draftState) => {
+      const permission = draftState.subjects[origin]?.permissions[target];
+      if (!permission) {
+        throw new PermissionDoesNotExistError(origin, target);
+      }
+
+      if (!permission.caveats) {
+        throw new PermissionHasNoCaveatsError(origin, target);
+      }
+
+      const caveatIndex = permission.caveats.findIndex(
+        (existingCaveat) => existingCaveat.type === caveatType,
+      );
+      if (caveatIndex === -1) {
+        throw new CaveatDoesNotExistError(origin, target, caveatType);
+      }
+
+      if (permission.caveats.length === 1) {
+        permission.caveats = null;
+      } else {
+        permission.caveats.splice(caveatIndex, 1);
+      }
+
+      // Typecast: ts(2589)
+      this.validatePermission(permission as any);
+    });
+  }
+
+  /**
+   * Used for retrieving the key that manages the restricted method
+   * associated with the current RPC `method` key.
+   *
+   * Used to support our namespaced method feature, which allows blocks
+   * of methods to be hidden behind a restricted method with a trailing `_` character.
```

Comment:
What feature is this? I'm having trouble following what this method is for.

On my first read through it seems like it takes an underscore-delimited method, splits it up, then puts it back together again. Except that it also checks for whether the method corresponds to a group of methods, in which case it returns the wildcard version.